### PR TITLE
Implement `too-many-positional-arguments`

### LINKED ISSUE #9099: [feature request] New warning: too many positional arguments
### Current problem

With R0913 too-many-arguments now including keyword arguments (#8667), the reasonable setting for `max-args` is now larger, since it's not really a problem to have a dozen keyword-only arguments.

However, having a dozen positional arguments remains a poor design.

### Desired solution

A new warning for too many positional arguments, with a separately configurable maximum. This would warn when there are more than a very small number of positional arguments (including positional-only arguments).

This would be in addition to the existing too-many-arguments warning.

### Additional context

This has been previously suggested in comments on #8667

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description
Closes #9099

### CODE REVIEW

File: doc/data/messages/t/too-many-positional/bad.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class FiveArgumentMethods:
+    """The max positional arguments default is 5."""
+
+    # +1: [too-many-arguments, too-many-positional]]
+    def take_five_args(self, a, b, c, d, e):
```
Comment (jacobtylerwalls) [inline doc/data/messages/t/too-many-positional/bad.py:5]: I welcome input on making this more user-friendly--I was just hammering this out quickly.

File: doc/user_guide/configuration/all-options.rst [hunk 2]
```diff
@@ -765,6 +765,13 @@ Standard Checkers
 **Default:**  ``7``
 
 
+--max-positional
+""""""""""""""""
+*Maximum number of positional arguments for function / method.*
+
+**Default:**  ``5``
```
Comment (jacobtylerwalls) [inline doc/user_guide/configuration/all-options.rst:772]: What should the default be? I've worked on projects where the limit was four, including self (three if static). 5 seems safe for now?

File: pylint/checkers/design_analysis.py [hunk 3]
```diff
@@ -81,11 +82,9 @@
             "Used when an if statement contains too many boolean expressions.",
         ),
         "R0917": (
-            "Too many positional arguments in a function call.",
+            "Too many positional arguments (%s/%s)",
             "too-many-positional",
```
Comment (jacobtylerwalls) [inline pylint/checkers/design_analysis.py:86]: OR: `too-many-positional-args`?

File: doc/user_guide/configuration/all-options.rst (hunk 2 above)
Comment (DanielNoord) [reply doc/user_guide/configuration/all-options.rst:772]: Although I feel like `--max-positional-arguments` makes more sense.
Most people will use a config file so the added verbosity is a cost only paid once and then actually makes the config file a little bit more understandable.

File: doc/whatsnew/fragments/9099.new_check [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Added `too-many-positional` to allow distinguishing the configuration for too many
+total arguments (with keyword-only params specified after `*`) from the configuration
+for too many positional-or-keyword or positional-only arguments.
+
+As part of evaluating whether this check makes sense for your project, ensure you
+adjust the value of `--max-positional`.
```
Comment (DanielNoord) [inline doc/whatsnew/fragments/9099.new_check:6]: See comment above about the option name

File: pylint/checkers/design_analysis.py (hunk 3 above)
Comment (DanielNoord) [reply pylint/checkers/design_analysis.py:86]: I'd prefer the rename yeah

File: pylintrc [hunk 5]
```diff
@@ -433,6 +433,9 @@ max-attributes=11
 # Maximum number of statements in a try-block
 max-try-statements = 7
 
+# Maximum number of positional arguments (see R0917).
+max-positional = 12
```
Comment (DanielNoord) [inline pylintrc:437]: 😂

File: doc/data/messages/t/too-many-positional/details.rst [hunk 6]
```diff
@@ -1 +0,0 @@
-Reserved message name, not yet implemented.
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/t/too-many-positional/details.rst:1]: Probably need some explanation about how it's hard to remember the order of everything if there's a lot of positional and thus it's a bad API. (We can take some inspiration from what Ruff says)

File: pylint/checkers/design_analysis.py [hunk 7]
```diff
@@ -81,11 +82,9 @@
             "Used when an if statement contains too many boolean expressions.",
         ),
         "R0917": (
-            "Too many positional arguments in a function call.",
-            "too-many-positional",
-            "Will be implemented in https://github.com/pylint-dev/pylint/issues/9099,"
-            "msgid/symbol pair reserved for compatibility with ruff, "
-            "see https://github.com/astral-sh/ruff/issues/8946.",
+            "Too many positional arguments (%s/%s)",
+            "too-many-positional-arguments",
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/design_analysis.py:86]: I also prefer the new name. But Ruff chose the other one and we originally created this message in anticipation for compatibility with ruff. Is it worth the possible confusion ?

File: pylintrc [hunk 8]
```diff
@@ -433,6 +433,9 @@ max-attributes=11
 # Maximum number of statements in a try-block
 max-try-statements = 7
 
+# Maximum number of positional arguments (see R0917).
+max-positional-arguments = 12
```
Comment (Pierre-Sassoulas) [inline pylintrc:437]: Oh, wow :smile:

File: doc/user_guide/configuration/all-options.rst (hunk 2 above)
Comment (Pierre-Sassoulas) [reply doc/user_guide/configuration/all-options.rst:772]: Yeah let's try to not annoy anyone too much. Afterall the pylint code base itself needs a 12 :D

**Review by Pierre-Sassoulas — COMMENTED** [review]
Great !

File: pylint/checkers/design_analysis.py (hunk 7 above)
Comment (DanielNoord) [reply pylint/checkers/design_analysis.py:86]: We can easily rename. `ruff` only supports codes for now, so renaming is just a change in output and documentation. That should be fine for them? Considering https://github.com/astral-sh/ruff/issues/8946 was merged by @charliermarsh perhaps they can confirm that a rename should be fine for them as well?

File: doc/user_guide/configuration/all-options.rst (hunk 2 above)
Comment (jacobtylerwalls) [reply doc/user_guide/configuration/all-options.rst:772]: Yeah it shouldn't be higher than max total arguments, which defaults to 5 also.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Review by DanielNoord — APPROVED** [review]

File: pylint/checkers/design_analysis.py (hunk 7 above)
Comment (Pierre-Sassoulas) [reply pylint/checkers/design_analysis.py:86]: They fixed it in 0.6.1 / https://github.com/astral-sh/ruff/pull/12905. There's a comment saying we should have named it ``"too-many-positional-parameters`` (https://github.com/astral-sh/ruff/issues/12619#issuecomment-2265121518).

Comment (jacobtylerwalls) [reply pylint/checkers/design_analysis.py:86]: I get it. But leaving it this way maintains the rhyme between the existing `max-args` setting and the new `max-positional-arguments` setting.

Comment (Pierre-Sassoulas) [reply pylint/checkers/design_analysis.py:86]: Yeah and the churn in ruff and pylint if we change it again would be pretty bad. (tbh I didn't consciously make this distinction between arguments/parameters internally myself)

**Pull request merged**